src/interpreter.py

```diff
@@ -1,143 +1,164 @@
-            parameters, body = func
-            if len(parameters) != len(expr.arguments):
-                raise Exception(f"Function '{expr.name}' expects {len(parameters)} arguments, got {len(expr.arguments)}")
-            arg_values = [self.evaluate(arg) for arg in expr.arguments]
-            func_env = Environment(parent=self.env)
-            for param, arg_value in zip(parameters, arg_values):
-                func_env.set(param, arg_value)
-            original_env = self.env
-            self.env = func_env
-            try:
-                for stmt in body:
-                    self.interpret(stmt)
-            except ReturnException as e:
-                return e.value
-            finally:
-                self.env = original_env
-            return None 
-
-        else:
-            raise Exception(f"Unsupported expression: {type(expr).__name__}")
-
-    def _apply_operator(self, operator, left, right):
-        if operator == "SUM OF":
-            return left + right
-        elif operator == "DIFF OF":
-            return left - right
-        elif operator == "PRODUKT OF":
-            return left * right
-        elif operator == "QUOSHUNT OF":
-            return left / right
-        elif operator == "MOD OF":
-            return left % right
-        elif operator == "BIGGR OF":
-            return max(left, right)
-        elif operator == "SMALLR OF":
-            return min(left, right)
-        elif operator == "BOTH OF":
-            return bool(left) and bool(right)
-        elif operator == "EITHER OF":
-            return bool(left) or bool(right)
-        elif operator == "WON OF":
-            return bool(left) != bool(right)
-        else:
-            raise Exception(f"Unknown operator: {operator}")
-        
\ No newline at end of file
+        
+        if isinstance(expr, BinaryOperation):
+            return self._evaluate_binary_operation(expr)
+        
+        if isinstance(expr, FunctionCall):
+            return self._call_function(expr)
+        
+        raise RuntimeError(f"Unsupported expression: {type(expr).__name__}")
+
+    def _execute_statements(self, statements: list) -> None:
+        """Execute a sequence of statements."""
+        for stmt in statements:
+            self.interpret(stmt)
+
+    def _handle_if_statement(self, node: IfStatement) -> None:
+        """Execute conditional branching logic."""
+        condition = self.evaluate(node.condition)
+        block = node.true_block if condition else node.false_block
+        if block:
+            self._execute_statements(block)
+
+    def _handle_loop(self, node: Loop) -> None:
+        """Execute loop iteration logic."""
+        start_val = self.evaluate(node.start_expr)
+        end_val = self.evaluate(node.end_expr)
+        
+        for i in range(start_val, end_val + 1):
+            self.env.set(node.variable_name, i)
+            self._execute_statements(node.body)
```

Comment:
Недостатній контроль типів у циклі.

```suggestion
    def _handle_loop(self, node: Loop) -> None:
        """Execute loop iteration logic."""
        start_val = self.evaluate(node.start_expr)
        end_val = self.evaluate(node.end_expr)
        
        if not isinstance(start_val, int) or not isinstance(end_val, int):
            raise RuntimeError("Loop bounds must be integers")
        
        for i in range(start_val, end_val + 1):
            self.env.set(node.variable_name, i)
            self._execute_statements(node.body)
```
Задля виправлення можна додати перевірку.